src/platforms/esp32/main/ledc_nifs.c

```diff
@@ -0,0 +1,503 @@
+    VALIDATE_VALUE(speed_mode, term_is_integer);
+    term channel = argv[1];
+    VALIDATE_VALUE(channel, term_is_integer);
+
+    uint32_t duty = ledc_get_duty(
+        term_to_int(speed_mode),
+        term_to_int(channel)
+    );
+    switch (duty) {
+        case LEDC_ERR_DUTY:
+            fprintf(stderr, "Unable to ledc_get_duty.\n");
+            RAISE_ERROR(term_from_int(LEDC_ERR_DUTY));
+        default:
+            TRACE("ledc_get_duty\n");
+            return term_from_int(duty);
+    }
+}
+
+static term nif_ledc_set_duty(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term speed_mode = argv[0];
+    VALIDATE_VALUE(speed_mode, term_is_integer);
+    term channel = argv[1];
+    VALIDATE_VALUE(channel, term_is_integer);
+    term duty = argv[2];
+    VALIDATE_VALUE(duty, term_is_integer);
+
+    esp_err_t err = ledc_set_duty(
+        term_to_int(speed_mode),
+        term_to_int(channel),
+        term_to_int(duty)
+    );
+    switch (err) {
+        case ESP_OK:
+            TRACE("ledc_set_duty\n");
+            return OK_ATOM;
+        default:
+            fprintf(stderr, "Unable to ledc_set_duty. err=%i\n", err);
+            RAISE_ERROR(term_from_int(err));
+    }
+}
+
+static term nif_ledc_update_duty(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term speed_mode = argv[0];
+    VALIDATE_VALUE(speed_mode, term_is_integer);
+    term channel = argv[1];
+    VALIDATE_VALUE(channel, term_is_integer);
+
+    esp_err_t err = ledc_update_duty(
+        term_to_int(speed_mode),
+        term_to_int(channel)
+    );
+    switch (err) {
+        case ESP_OK:
+            TRACE("ledc_update_duty\n");
+            return OK_ATOM;
+        default:
+            fprintf(stderr, "Unable to ledc_update_duty. err=%i\n", err);
+            RAISE_ERROR(term_from_int(err));
+    }
+}
+
+static term nif_ledc_get_freq(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term speed_mode = argv[0];
+    VALIDATE_VALUE(speed_mode, term_is_integer);
+    term timer_num = argv[1];
+    VALIDATE_VALUE(timer_num, term_is_integer);
+
+    uint32_t freq_hz = ledc_get_freq(
+        term_to_int(speed_mode),
+        term_to_int(timer_num)
+    );
+    switch (freq_hz) {
+        case 0:
+            fprintf(stderr, "Unable to ledc_get_freq.\n");
+            RAISE_ERROR(term_from_int(0));
+        default:
+            TRACE("ledc_get_freq\n");
+            return term_from_int(freq_hz);
+    }
+}
+
+static term nif_ledc_set_freq(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+    term speed_mode = argv[0];
+    VALIDATE_VALUE(speed_mode, term_is_integer);
+    term timer_num = argv[1];
+    VALIDATE_VALUE(timer_num, term_is_integer);
+    term freq_hz = argv[2];
+    VALIDATE_VALUE(freq_hz, term_is_integer);
+
+    esp_err_t err = ledc_set_freq(
+        term_to_int(speed_mode),
+        term_to_int(timer_num),
+        term_to_int(freq_hz)
+    );
+    switch (err) {
```

Comment:
Using  switch here just for 2 cases is a bit an overkill, I would just do:
```
if (UNLIKELY(res != ESP_OK)) {
    fprintf(stderr, "Unable to ledc_set_freq. err=%i\n", err);
    RAISE_ERROR(term_from_int(err));
}
```

Which is way more comapct.

Reply:
I changed all the functions that check return values to the pattern you suggest.